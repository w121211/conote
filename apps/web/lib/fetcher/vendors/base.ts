import got, { HTTPError } from 'got'
// import createMetascraper from 'metascraper'
// import metascraperAuthor from 'metascraper-author'
// import metascraperDate from 'metascraper-date'
// import metascraperDescription from 'metascraper-description'
// import metascraperLang from 'metascraper-lang'
// import metascraperTitle from 'metascraper-title'
// import metascraperUrl from 'metascraper-url'
import { createScraper } from '@conote/scraper'
import yfinancePack from '@conote/scraper/dist/packs/yfinance'
import keywordsPack from '@conote/scraper/dist/packs/keywords'
import tickersPack from '@conote/scraper/dist/packs/tickers'
import type { DomainFetchFunction } from './index'

// const metascraper = createMetascraper([
//   metascraperAuthor(),
//   metascraperDate(),
//   metascraperDescription(),
//   metascraperLang(),
//   metascraperTitle(),
//   metascraperUrl(),
// ])
const metascraper = require('metascraper')([
  require('metascraper-author')(),
  require('metascraper-date')(),
  require('metascraper-description')(),
  // require('metascraper-image')(),
  require('metascraper-lang')(),
  // require('metascraper-clearbit')(),
  // require('metascraper-publisher')(),
  require('metascraper-title')(),
  require('metascraper-url')(),
])

const extraScraper = createScraper([yfinancePack, keywordsPack, tickersPack])

export type ParseUrlResult = {
  resolvedUrl: string
  domain: string
}

export function parseUrl(url: string): ParseUrlResult {
  // TODO: 不同的URL(eg: short-url)可能指向同一頁面，需整合 特別是URL params
  const u = new URL(url)
  return {
    domain: u.hostname,
    resolvedUrl: url,
  }
}

export const base: DomainFetchFunction = async url => {
  try {
    const { body: html, url: finalUrl } = await got(url)
    const {
      url: scrapedUrl,
      author,
      date,
      description,
      lang,
      title,
    } = await metascraper({ html, url: finalUrl })
    const { keywords, tickers } = await extraScraper({ html, url: finalUrl })
    const { domain } = parseUrl(finalUrl)
    // if (domain === undefined) {
    //   throw new Error(`Fetch error: ${url} ${finalUrl}`)
    // }

    return {
      domain,
      finalUrl: scrapedUrl,
      srcType: 'other',
      authorName: author,
      date,
      description,
      lang,
      title,
      keywords,
      tickers,
    }
  } catch (err) {
    if (err instanceof HTTPError) {
      const { domain } = parseUrl(url)
      return {
        domain,
        finalUrl: url,
        srcType: 'other',
        error: err.message,
      }
    } else {
      throw err
    }
  }
}
