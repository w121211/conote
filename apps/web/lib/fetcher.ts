import { resolve } from 'path'
import { FetchClient } from './fetcher/fetcher'

const fetchClientPropertyName = `__prevent-name-collision__fetcher`

type GlobalThisWithFetchClient = typeof globalThis & {
  [fetchClientPropertyName]: FetchClient
}

function getFetchClient(): FetchClient {
  if (process.env.NODE_ENV === `production`) {
    return new FetchClient()
  } else {
    const newGlobalThis = globalThis as GlobalThisWithFetchClient
    if (!newGlobalThis[fetchClientPropertyName]) {
      newGlobalThis[fetchClientPropertyName] = new FetchClient(resolve(process.cwd(), '.fetcher_local_cache.dump.json'))
    }
    return newGlobalThis[fetchClientPropertyName]
  }
}

const fetcher = getFetchClient()

export default fetcher
