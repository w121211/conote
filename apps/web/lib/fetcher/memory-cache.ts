/**
 * A simple memory cache, mainly use for avoiding repeatly calls on the same urls during development
 */
import { readFileSync, writeFileSync } from 'fs'
import { LinkScrapedResult } from '../interfaces'

export class MemoryCache {
  private filepath: string
  private cache: Record<string, LinkScrapedResult> = {}

  constructor(cachePath: string) {
    this.filepath = cachePath
    try {
      this.cache = this.load(this.filepath)
    } catch (err) {
      console.log(`cache not found, create a new one`)
      //   console.log('No ')
    }
  }

  private load(filepath: string): Record<string, LinkScrapedResult> {
    console.log(`loading cache from ${filepath}`)
    return JSON.parse(readFileSync(filepath, { encoding: 'utf8' }))
  }

  public dump(): void {
    console.log(`Dump cache to ${this.filepath}`)
    writeFileSync(this.filepath, JSON.stringify(this.cache))
  }

  public get(url: string): LinkScrapedResult {
    if (url in this.cache) {
      console.log('hit cache: ' + url)
      return this.cache[url]
    }
    throw new Error('Cache not found:' + url)
  }

  public set(url: string, result: LinkScrapedResult): void {
    this.cache[url] = result
  }
}
