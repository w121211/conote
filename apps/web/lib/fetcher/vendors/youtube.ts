import { google } from 'googleapis'
import { LinkScrapedResult } from '../../interfaces'
import { DomainFetchFunction, DomainNotFitError } from './index'

const reYoutube = {
  video_id: [
    /(?:http[s]?:\/\/)?(?:\w+\.)?youtube.com\/watch\?v=([\w_-]+)(?:&.*)?/i,
    /(?:http[s]?:\/\/)?youtu.be\/([\w_-]+)(?:\?.*)?/i,
  ],
  playlist_id: [
    /(?:http[s]?:\/\/)?(?:\w+\.)?youtube.com\/playlist\?list=([\w_-]+)(?:&.*)?/i,
  ],
  channel_user: [
    /(?:http[s]?:\/\/)?(?:\w+\.)?youtube.com\/user\/([\w_-]+)(?:\?.*)?/i,
  ],
  channel_id: [
    /(?:http[s]?:\/\/)?(?:\w+\.)?youtube.com\/channel\/([\w_-]+)(?:\?.*)?/i,
  ],
  channel_custom: [
    /(?:http[s]?:\/\/)?(?:\w+\.)?youtube.com\/c\/([\w_-]+)(?:\?.*)?/i,
    /(?:http[s]?:\/\/)?(?:\w+\.)?youtube.com\/([\w_-]+)(?:\?.*)?/i,
  ],
}

const youtubeApi = google.youtube({
  version: 'v3',
  auth: process.env.YOUTUBE_API_KEY,
})

function parseUrl(url: string): { vid: string } {
  for (const e of reYoutube.video_id) {
    const result = e.exec(url)
    if (result) {
      return { vid: result[1] }
    }
  }
  throw new DomainNotFitError(`Not found youtube video id: ${url}`)
}

export const youtube: DomainFetchFunction = async url => {
  const { vid } = parseUrl(url) // include test domain

  const resp = await youtubeApi.videos.list({
    part: ['snippet'],
    id: [vid],
  })
  if (resp.data.items && resp.data.items[0]) {
    const e = resp.data.items[0]
    const res: LinkScrapedResult = {
      domain: 'youtube.com',
      finalUrl: url, // TODO: (bug) possibly be redirect, short-url, mobile-url
      srcType: 'video',
      srcId: vid,
      authorId: e.snippet?.channelId ?? undefined,
      authorName: e.snippet?.channelTitle ?? undefined,
      date: e.snippet?.publishedAt
        ? new Date(e.snippet.publishedAt).toISOString()
        : undefined,
      description: e.snippet?.description ?? undefined,
      title: e.snippet?.title ?? undefined,
      keywords: e.snippet?.tags ?? undefined,
    }
    return res
  }

  throw new Error(`Youtube fetcher failed: ${url}`)
}
