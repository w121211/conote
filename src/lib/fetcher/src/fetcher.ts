import got from 'got'
// import KeyvRedis from '@keyv/redis'
import { FetchResult, SrcType } from './index'

// const keyv = new KeyvRedis('redis://redis:6379')
// keyv.on('error', (err: any) => console.log('Connection Error', err))

export const fetcher: Record<string, (url: string, domain?: string) => Promise<FetchResult>> = {
  youtube: async function (url: string) {
    const domain = 'youtube'
    const u = new URL(url)
    const vid = new URLSearchParams(u.search).get('v')
    // try {
    // const resp = await got.get('https://youtube.com/get_video_info', { searchParams: { video_id: vid } })
    // } catch (error) {
    //     console.log(error.response.body)
    // }
    const resp = await got.get('https://youtube.com/get_video_info', {
      searchParams: { video_id: vid },
      // cache: keyv,
    })
    const p = new URLSearchParams(resp.body).get('player_response')
    if (p) {
      const j = JSON.parse(p)
      return {
        domain,
        // TODO: 可能會有redirect, short-url
        resolvedUrl: url,
        srcType: SrcType.VIDEO,
        srcId: vid ?? '',
        srcTitle: j.microformat.playerMicroformatRenderer.title.simpleText,
        srcPublishDate: new Date(j.microformat.playerMicroformatRenderer.publishDate).toISOString(),
        authorId: j.microformat.playerMicroformatRenderer.externalChannelId,
        authorName: j.microformat.playerMicroformatRenderer.ownerChannelName,
        keywords: j.videoDetails.keywords,
        description: j.videoDetails.shortDescription,
      }
    }
    throw new Error(`youtube fetcher failed: ${url}`)
  },
  default: async function (url, domain) {
    if (domain === undefined) throw new Error()
    return {
      domain,
      resolvedUrl: url,
      srcType: SrcType.OTHER,
    }
  },
}

// function youtube(): void {
//   console.log('hi')
// }

// const fetchers = [youtube]

// function fetch() {
//   const domain = resolveDomain(url)
//   const fetcher = fetchers.find(e => e.matcher(domain))
//   fetcher
// }
