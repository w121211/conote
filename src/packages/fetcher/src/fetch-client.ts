import { FetchResult } from './index'
import { LocalCache } from './local-cache'
import { tryFetch } from './vendors/index'

export class FetchClient {
  // 有cache支援
  private cache: LocalCache | null = null

  constructor(localCachePath: string | null = null) {
    if (localCachePath) {
      this.cache = new LocalCache(localCachePath)
    }
  }

  public async fetch(url: string): Promise<FetchResult & { fromCache?: true }> {
    if (this.cache) {
      try {
        return { ...this.cache.get(url), fromCache: true }
      } catch {
        const res = await tryFetch(url)
        this.cache.set(url, res)
        return res
      }
    }
    return await tryFetch(url)
  }

  public dump(): void {
    if (this.cache) {
      this.cache.dump()
    }
  }
}
