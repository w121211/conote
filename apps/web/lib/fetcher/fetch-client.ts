import { MemoryCache } from './memory-cache'
import { tryFetch } from './vendors/index'

export type SrcType = 'VIDEO' | 'POST' | 'AUTHOR' | 'OTHER'

export type FetchResult = {
  domain: string
  finalUrl: string
  srcId?: string
  srcType: SrcType

  // metascraper
  authorId?: string
  authorName?: string
  date?: string
  description?: string
  lang?: string
  title?: string

  // ./packages/scraper
  keywords?: string[]
  tickers?: string[]
  error?: string
}

export class FetchClient {
  private cache: MemoryCache | null = null // cache support

  constructor(localCachePath: string | null = null) {
    if (localCachePath) {
      this.cache = new MemoryCache(localCachePath)
    }
  }

  public async fetch(url: string): Promise<FetchResult & { fromCache?: true }> {
    if (this.cache) {
      try {
        return {
          ...this.cache.get(url),
          fromCache: true,
        }
      } catch {
        const res = await tryFetch(url)
        this.cache.set(url, res)
        return res
      }
    }
    return await tryFetch(url)
  }

  public dump(): void {
    if (this.cache) {
      this.cache.dump()
    }
  }
}
