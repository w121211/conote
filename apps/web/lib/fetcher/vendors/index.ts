import { base } from './base'
import { youtube } from './youtube'
import { FetchResult } from '../fetch-client'

export type DomainFetchFunction = (url: string, domain?: string) => Promise<FetchResult>

export class DomainNotFitError extends Error {
  constructor(message?: string) {
    super(message) // 'Error' breaks prototype chain here
    Object.setPrototypeOf(this, DomainNotFitError.prototype) // restore prototype chain
  }
}

// fallback 依序嘗試執行，若失敗則嘗試下一個
const fetchers: DomainFetchFunction[] = [youtube, base]

export async function tryFetch(url: string): Promise<FetchResult> {
  for (const fetcher of fetchers) {
    try {
      // eslint-disable-next-line no-await-in-loop
      return await fetcher(url)
    } catch (err) {
      if (err instanceof DomainNotFitError) {
        continue
      } else {
        console.error(err)
        break
      }
    }
  }
  throw new Error(`Fetch error: ${url}`)
}
