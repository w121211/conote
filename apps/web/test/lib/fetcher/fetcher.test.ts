import { readFileSync, unlinkSync } from 'fs'
import { resolve } from 'path'
import { FetchClient } from '../../../lib/fetcher/fetch-client'
import { tryFetch } from '../../../lib/fetcher/vendors/index'

const failUrls = [
  'https://www.mobile01.com/topicdetail.php?f=803&t=6541514', // fail only on server, possibly caused by cloudflare guard
]

describe('tryFetch()', () => {
  it('fetch youtube', async () => {
    expect(await tryFetch('https://www.youtube.com/watch?v=CKljNQ5xe2w')).toMatchSnapshot()
  })

  it('fetch general', async () => {
    expect(await tryFetch('https://test.com')).toMatchSnapshot()
  })

  it.each(failUrls)('failUrl %s', async url => {
    expect(await tryFetch(url)).toMatchSnapshot()
  })
})

describe('class FetchClient', () => {
  const localCachePath = resolve(process.cwd(), './localcache.test.json')
  let fetcher: FetchClient

  // 這個無法使用，可能是因為fetcher是class
  // jest.spyOn(index, 'fetch').mockImplementation(async () => {
  //   console.log('fetch~~~~~')
  //   return {
  //     domain: 'mock.test.com',
  //     resolvedUrl: 'https://mock.test.com',
  //     srcType: 'OTHER',
  //   }
  // })
  // expect(fetchSpy).toBeCalled()

  beforeAll(async () => {
    fetcher = new FetchClient(localCachePath)
    await fetcher.fetch('https://test.com')
  })

  afterAll(() => {
    unlinkSync(localCachePath)
  })

  it('use cache', async () => {
    expect(await fetcher.fetch('https://test.com')).toEqual({
      fromCache: true,
      domain: 'test.com',
      resolvedUrl: 'https://test.com',
      srcType: 'OTHER',
    })
  })

  it('dump cache', async () => {
    fetcher.dump()
    expect(JSON.parse(readFileSync(localCachePath, { encoding: 'utf8' }))).toEqual({
      'https://test.com': { domain: 'test.com', resolvedUrl: 'https://test.com', srcType: 'OTHER' },
    })
  })
})
