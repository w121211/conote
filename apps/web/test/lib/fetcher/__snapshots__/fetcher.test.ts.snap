// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tryFetch() failUrl https://www.mobile01.com/topicdetail.php?f=803&t=6541514 1`] = `
Object {
  "authorName": "Evelyn Tsai",
  "date": "2022-02-17T09:33:55.000Z",
  "description": "傳言甚囂塵上已久，現如今LINE Pay與一卡通正式分家，LINE Pay App在這一個星期二（2022年2月15日）的時候更新版本，正式切割一卡通電子支付，約有高達四百五十五萬一卡通電子支付的用戶未來將會無法再透過LINE Pay App使用服務。LINE Pay自從西元2015...(行動支付 第1頁)",
  "domain": "www.mobile01.com",
  "finalUrl": "https://www.mobile01.com/topicdetail.php?f=803&t=6541514",
  "keywords": Array [
    "行動支付",
    "理財",
  ],
  "lang": "zh",
  "srcType": "OTHER",
  "tickers": Array [
    "LINE",
    "APP",
  ],
  "title": "LINE Pay鐵了心正式切割一卡通！455萬用戶會淪為電支棄兒嗎？ - Mobile01",
}
`;

exports[`tryFetch() fetch youtube 1`] = `
Object {
  "authorId": "UCACB9uLz5-62dEfpd8c5m7A",
  "authorName": "YumYum얌얌",
  "date": "2021-05-19T03:00:03.000Z",
  "description": "#한식 #몰아보기
💗
00:00 대구 마을불고기 돼지불백 정식
https://goo.gl/maps/RuNUo6a4U59nFWjC7
11:13 담양 삼거리 농원 닭볶음탕
https://goo.gl/maps/omjY4W6RWjuAHs1r5
18:21 수원 지동시장 올레 쭈꾸미
https://goo.gl/maps/gdZiBRct6gHQSHC16
23:11 구미 성주순대
https://goo.gl/maps/s61Eax3tXottBmwp6
31:16 석계역 서진박사 곱창마을
https://goo.gl/maps/aA3jxQW5HKoUAFfSA
37:59 대구 서문시장 약초골목 칼국수
https://goo.gl/maps/aXj4GKQwCMXj9eWa6
45:12 서울 창동 즐거운 술상
https://goo.gl/maps/1B4XRuCfyaWcDhkt6

💗Social media
▶인스타그램 : https://bit.ly/39D5adW
▶네이버블로그 : https://bit.ly/2VQ3mdA

💗비즈니스 문의

▶akongfood@gmail.com

Copyright 2020. Yum Yum All Videos Cannot Be Copied Without Permission",
  "domain": "youtube.com",
  "finalUrl": "https://www.youtube.com/watch?v=CKljNQ5xe2w",
  "keywords": Array [
    "한국음식",
    "길거리음식",
    "Korean street food",
    "street food",
    "yum yum",
    "korean food",
    "japanese food",
    "만들기",
    "수제",
    "recipes",
    "seafood",
    "맛집 추천",
    "food",
    "kroean rice soup",
    "cooking",
    "homemade",
    "kpop",
    "spicy food",
    "korean seafood",
    "korean recipes",
    "korean chicken",
    "김치음식",
    "요리레시피",
    "street of food",
    "street food market",
    "recipe to",
    "snacks",
    "spicy korean rice",
    "korean fried food",
    "korean food in korea",
    "생활의달인",
    "레시피영상",
    "best street foods",
    "Koreanseafood",
    "kfood",
    "얌얌",
    "yumyum얌얌",
    "korean street food",
    "돼지불백",
    "이모카세",
    "수제비",
    "쭈꾸미",
    "순대국밥",
    "곱창",
    "닭볶음탕",
    "pork",
    "sujebi",
    "sundae",
    "chicken",
  ],
  "srcId": "CKljNQ5xe2w",
  "srcType": "VIDEO",
  "title": "이런게 바로 한식 ! 돼지 불백, 닭볶음탕, 순대국밥, 칼국수, 쭈꾸미 | A collection of various Korean foods | Korean Street food",
}
`;
